README.md
HopperGuidance v0.2.3
=====================

A KSP mod which enables the calculation of a fuel efficient trajectory in-flight through a number of targets to make a soft landing at a designated location respecting the capabilities of the craft. The craft will be steered automonously to follow this trajectory. You can choose many parameters of the trajectory to make it safe or more exciting. This has been a culmination of years of effort on and off, particularly to give to the ability to find the optimum path to steer quickly through a sequence of targets.

Features
- Pick targets to pass through ending on a ground target
- Shows the trajectory to landing and required thrust vectors
- Lets you tune the parameters to the craft

HopperGuidance is designed handle trajectories close to the landing site but you can see how far you can push it. It may lose accuracy beyond 10km. It can't handle descent from orbit well as it treats the planet surface as flat and doesn't take into account atmospheric drag, though can correct for it while flying. It may waste quite a bit of fuel trying to keep to the pre-calculated trajectory too. You can give it a try though, its all good fun.

The algorithm used is aimed to be an implementation of the G-FOLD algorithm for fuel optimal diverts
reputably used by Lars Blackmore for landing the SpaceX Falcon-9 rocket. However my algorithm is many ways a simplified version of it, though I've changed the constraints quite a bit and have extended it to compute trajectories though intermediate targets which involved some compromises, but it works pretty well.

Please see the videos from earlier versions here https://www.youtube.com/watch?v=Ekywp-P6EBQ&t=301
and https://www.youtube.com/watch?v=ouIQUfHsZac&t=101s
For the latest version, which is much improved, see https://www.youtube.com/watch?v=unJmEusvYHs

References

-  "G-FOLD: A Real-Time Implementable Fuel Optimal Large Divert Guidance Algorithm for Planetary Pinpoint Landing" https://www.researchgate.net/publication/258676350_G-FOLD_A_Real-Time_Implementable_Fuel_Optimal_Large_Divert_Guidance_Algorithm_for_Planetary_Pinpoint_Landing.

- "Convex Programming Approach to Powered Descent Guidance for Mars Landing" https://arc.aiaa.org/doi/abs/10.2514/1.27553

- "Minimum-Landing-Error Powered-Descent Guidance for Mars Landing Using Convex Optimization" http://www.larsblackmore.com/BlackmoreEtAlJGCD10.pdf

Pre-requistes
=============

Works with KSP versions 1.8.1 to 1.10 and possibly beyond.
If you want to run the solver outside of KSP you need python and matplotlib to show the trajectories.

Installing
==========

Unpack HopperGuidance-v0.2.3.zip (from releases) into your GameData folder inside your KSP installation.

Using HopperGuidance
====================

The mod includes shiny cuboid HopperGuidanceCore part in the Command & Control category and two test craft, Test Craft and Test Craft Heavy. Right click to the part to show its UI.

Setting the target position. Click "Pick target" to select a ground position with the mouse pointer.
You can changes its height via the part dialog box which always sets the height of the last target.
A yellow target, inspired a little by the SpaceX drone ship markings will be shown. If the height is greater than zero then the centre cross is extended on a stick above the ground.

Once in flight click "Enable guidance" to automatically calculate a trajectory to safely land at the target point. A trajectory will be shown in green with the calculated thrust direction and magnitude in red. The craft will be steered to attempt to follow this trajectory. You may get the message "Failed: try again" in which case the craft is likely to be either: too far away, beyond max velocity or below the minimum descent angle. If you tweak these numbers the trajectory will be automatically recalculated. However, if the solution has failed because is the position of the craft, may be its too low than you will need to manually press click "Enable guidance" again.

You can cancel the autopilot with "Cancel guidance" and take control again. You might want to do this to recalculate the trajectory if the craft has diverged too much from it.

Control parameters. The algorithm is in two parts.

1. Calculating a trajectory
2. Following the trajectory

Parameters for (1) are used to compute the trajectory and can be used to control how 'tame' the trajectory is or how exciting is. If you want your craft to come slowly and gently down from above like Blue Origin you can choose the slightly more exciting Space X Falcon 9 landing, or more extreme doomed to failure trajectories coming in fast and very low.

- Min. descent angle - imagine an upside down cone with the apex at the landing point. The cone describes the safe area of descent. A high angle to the ground means the craft must descent steeply, only getting near the ground at the landing point.
- Max velocity - Keep velocity below this limit
- Max thrust angle - the maximum allow angle the craft will tilt. If 90 degrees then the craft can tip from vertical to the horizontal position but no more. If beyond >90 this value is ignored completely.

Click "Pick Target" then click on the main view to select the target, and then sets if height on the part UI if you wish. You will need to increase the target size to see far away targets. Note that the target is horizontal so it will get hidden on slopes.

Parameters for (2) describe what to do when the craft is off the trajectory. The nearest point on the trajectory is marked by a blue line from the craft to the trajectory. This nearest point takes into account and position and velocity with a little more weight for velocity. This makes the craft behave more smoothly rather than blindly aiming for the nearest point in position. So the craft tries to match the position and velocity of the nearest point. A correct to the thrust vector is calculated minimise the discrepancy. Six PID controllers are used to achieve this. 3 for the X,Y and Z positions and 3 for X, Y and Z velocities.

- Idle attitude angle - If the craft is pointing more this angle away from the required direction of thrust then idle the engine at 1% thrust. This prevent the craft thrusting in the wrong direction and waits of it to point correctly
- Correction factor - Sets the velocity to aim for to close the position error. If set of 1 then when 1m aware aim for a velocity of 1m/s towards the target. 0.15 to 0.25 are good values. Use down to 0.15 for large vessels that take longer to change their attitude. Lower this if the craft oscillates around the trajectory.

Finally there are a few toggle switches

- Keep engine ignited - Use in Realism Overhaul where engines have limited ignition and take time so start up
- Show track - Display track, align and steer vectors
- Logging - If enabled the files vessel.dat, target.dat and solution.dat are logged to the same directory as KSP.log

Here's a diagram to explain some of the parameters

![](docs/Diagram.png)

General tips:

- Have sufficient RCS or reaction wheel for your craft so it can change its attitude quickly enough (look at whether the steer vector is often misaligned with the crafts attitude)
- Enable RCS
- It can take longer to compute highly constrained trajectories for heavy craft or not find a solution at all. Retry, try something easier or reduce the number of targets.
- If your craft comes in too low to land either reduce the thrust angle to land more vertically or increase min descent angle
- If your craft is way off the trajectory its best to cancel and re-enable guidance again to compute a new trajectory

Failure to find solution:

Usually caused by one of the following

- The craft is going faster than max velocity
- The craft is below the minimum descent angle (to final target)
- You have run out of fuel
- Its impossible not to hit the ground
- The targets are too close and near impossible to fly between
- There are too many targets. About 10 is the maximum
- You have run out of ignitions (Realism Overhaul)
- Other. Sadly sometimes the solver just fails to find a solution. Simplify things or just try again

Fun things to try
=================

Position targets for a slalom course around some buildings
Find the bridges at KSC and try to fly under them
Try and reproduce the SN5 Starship hop in Realism Overhaul
Try and reproduce a Falcon 9 landing in Realism Overhaul
Try setting targets round and round in a circle to make a loop

Dealing with Small Craft
========================

Small craft are exciting, like a sports car.
You can try and push of max velocity, max thrust angle to high values. The craft will then be able to move quickly and accelerate quickly leads to exciting but possibly risky landings.

Dealing with Large Craft
========================

Large craft are hard to maneuver since they have high mass and will take time to change its attitude. This leads to it pointing is the wrong direction and either thrusting in that wrong direction or waiting to change its attitude. Both these mean it cannot following the track. Another problem is that atmospheric drag is not included when solving for the best trajectory. So here's some tips for dealing with large craft

- Keep max thrust angle low, says <15 degrees as this will keep the craft more upright and mean the craft is never pointing in the wrong direction for long
- Lower correction factor down to as low as 0.15. This will mean if the craft is off trajectory it will only move slowly towards the correct trajectory. If this value is too high it will overshoot. Observe whether the red steer attitude is far from the actual crafts attitude
- You can compensate for the lack of ability to steer the craft by adding more RCS thrusters or overly powerful reaction wheels
- Don't position targets too close to each other, especially if large velocity changes are required

Dealing with Realism Overhaul:

- Choose an engine that is throttleable
- Switch on keepIgnited otherwise you may use up all your ignitions very quickly
- Note that the engine has a minimum thrust and if this results in craft rising then you have too much thrust to hover. So choose your engine carefully
- Reaction wheels don't work too well and RCS might be limited. Thrust vectoring of the main engine will be needed to steer the craft, but that can work well if the engine is always ignited and has enough thrust

Hints for a Falcon 9 landing:

HopperGuidance will only use the currently activated engines, as it can't read the min/max thrust otherwise. With engines that can't throttle to 0% this mean they will be providing thrust which is a nuisance. To emulate a Falcon 9 landing as in the latest video I shutdown the engines while falling and enable them at minimum thrust at 2500m with a brief press of SHIFT. This only slows the rocket at about 2m/s/s and then click "Enable Guidance". Also note that with 3 merlin engines the minimum thrust is greater than gravity so a Falcon 9 can't hover or descend slowly with three engines. If you find you are hovering above the ground you will need to cancel guidance and cut the engine otherwise you will rise again. In reality SpaceX only use a single engine for the final part of the descent, shutting down the other two, and I plan to implement this in a future version.

Debugging
=========

If you have a problem look in the KSP directory for KSP.log. Log lines are proceeded by [HopperGuidance]
You will be able to find all the parameters when either a solution was requested and found or it failed. You can rerun by taking the parameters from the log line and using them directly on the command line of Solve.exe outside of KSP. See below.

Testing
=======

There is a way to run the algorithm outside of KSP via Solve.exe.
I'm running on the Mac so I run
```
mono ./Solve.exe r0=[-340,100,20] v0=[10,10,10] target=[-200,120,0] target=[150,50,0] rf=[0,0,0] vf=[0,0,0] g=9.8 maxThrustsBetweenTargets=2 > solution.dat; ./plotXYZ.py solution.dat
```
for example and this produces a data file with a trajectory.
This can be plotted (via matplotlib) by running

```
./plotXYZ.py --square solution.dat

sage: plotXYZ.py [-h] [--xmin XMIN] [--xmax XMAX] [--ymin YMIN] [--ymax YMAX]
                  [--zmin ZMIN] [--zmax ZMAX] [--vxmin VXMIN] [--vxmax VXMAX]
                  [--vymin VYMIN] [--vymax VYMAX] [--vzmin VZMIN]
                  [--vzmax VZMAX] [--tmax TMAX] [--amult AMULT] [--square]
                  [--showchecks]
                  filename [filename ...]

Plot vessel data logs (or solutions) with X,Y,Z,VX,VY,VZ and Throttle in
multiple plots

positional arguments:
  filename       Filename of TAB-separated data file, first line contains
                 column names. Should contain time,x,y,z,vx,vy,vz,ax,ay,ax

optional arguments:
  -h, --help     show this help message and exit
  --xmin XMIN    Minimum x position
  --xmax XMAX    Maximum x position
  --ymin YMIN    Minimum y position
  --ymax YMAX    Maximum y position
  --zmin ZMIN    Minimum z position
  --zmax ZMAX    Maximum z position
  --vxmin VXMIN  Minimum vx position
  --vxmax VXMAX  Maximum vx position
  --vymin VYMIN  Minimum vy position
  --vymax VYMAX  Maximum vy position
  --vzmin VZMIN  Minimum vz position
  --vzmax VZMAX  Maximum vz position
  --tmax TMAX    Maximum time
  --amult AMULT  Multiplier for scale up thrust acceleration lines
  --square       Make XY plot square (roughly as depends on window size)
  --showchecks   Show time checks for max vel. and min descent angle
```

![](docs/plotXYZ.png)

Note: The maximum acceleration is shown as the dotted line on the time vs mag(accel) plot and this is exceeded temporarily. The craft will not be able to produce this much acceleration and the craft with diverge from the trajectory. This is because the constraint is actually applied independently to each X, Y, Z axis so the total magnitude can be larger than the axis limit. This is a simplification to help speed up solving time but there is scope for improvement.

You can also run all the tests by running
```
./run_tests.sh
and
./run_test2.sh # multi-target tests
```
This is all designed to run on a Mac where you have a bash shell. It should be failing easy to convert though.

Logging
=======

If you switch on logging in the part UI window then the file vessel.dat and solution.dat with by logged to the KSP directory. solution.dat is what would be generated by Solve.exe but running inside the Game. vessel.dat is a log of the actual vessel parameters while trying to follow the trajectory.

Heres an example:

![](docs/logging.png)

You can see the vessel followed trajectory reasonably well though couldn't make the turn half way through. The vessel is Test Craft Heavy and the problem is that its attitude could not match that required quickly enough so it didn't slowdown faster enough. Look at the attitude error plot.

Compilation
===========

Running on a Mac I used Mono to compile C#.
I couldn't get the Mac Visual Studio to use the correct version of Mono for KSP.
But I did manage to get it to work with the Makefile.
Just type
```
make
```
But I include the Visual Studio project as it may be easy to get running on a Windows machine or on a Mac with some configuration hacks. Any help on this much appreciated.

ALGLIB
======

I make use of the free version of ALGLIB for convex optimisation. I use the free version provided until the GPL. Here's a description of ALGLIB from alglib.net

ALGLIB is a cross-platform numerical analysis and data processing library. It supports several programming languages (C++, C#, Delphi) and several operating systems (Windows and POSIX, including Linux). ALGLIB features include:

- Data analysis (classification/regression, statistics)
- Optimization and nonlinear solvers
- Interpolation and linear/nonlinear least-squares fitting
- Linear algebra (direct algorithms, EVD/SVD), direct and iterative linear solvers
- Fast Fourier Transform and many other algorithms
- ALGLIB Project offers you several editions of ALGLIB:

Details of the Algorithm
========================

My implementation is not quite the same as the 'proper' G-FOLD algorithm. I made some simplifications due to the convex solver code I had available and in some cases due to my lack of understanding.
The full solution makes account of the consumption of fuel and the improved acceleration when mass drops due to this. Mine does not. If you run out of fuel then sorry.

The proper G-FOLD uses a Second Order Cone Programming (SOCP) solver. This is fancy stuff, but basically it means that various constraints such a thrust directions and minimum descent angle are all cones. You can see that they are since they are a maximum angle from the vertical in all directions.
I couldn't find a SOCP solver but I could find a Quadratic Programming (QP) solver in the shape of alglib. See alglib.net. This means I use 4 planes rather than a cone so its an approximation that you probably won't notice.

To follow the trajectory I use 6 PID controllers. There are a pair of controller for each X, Y and Z. The controllers are very simple using only the proportional element, kP. The 1st controller say for the X axis sets a target (or setpoint) for the velocity, VX given the error in the X position. To start with we find the closest position on the trajectory by considering both position and velocity of the craft and trajectory and finding the nearest point. kP is given by "Correction factor". To this we add the actual desired velocity from the nearest point on the trajectory. So this says aim to move towards the X position at this speed while also moving along the trajectory. Then the second PID for X set the acceleration in the X direction given the mismatch between the velocity desired given the target velocity and the current velocity. kP in this controller is set from "Acceleration gain". This is done for the three dimensions and gives a desired correction thrust vector. This correction is added to the thrust vector already computed from the solution. Finally the thrust vector in limited by the max thrust angle from vertical which will prevent the craft leaning too much. The direction of this thrust vector is shown by the red bar drawn from the centre of the craft.

The recent changes in v0.2.3 was to enable multiple targets. Each target has a position whereas the final target also has a velocity constraint of a small vertical descent. The G-FOLD algoritm has to test out solutions over a range of durations with a golden search. But with multiple targets there are multiples times so the search would quickly become high dimensional and take for other. To get around this the solution is done iteratively, computing the solution to the first target with minimum fuel and then pinning down the time of the first target. Then the next target is adding, trying out a range of durations to travel to it, and continuing. The trajectory from the start point to target 1 will generally needs changing once target 2 is known so the whole trajectory is computed again as each target is added, but there is a problem if target 2 requires a radical change of direction it might mean the craft can't get to target 1 so quickly, so a slight extra time allowance is the made to get to each target to allow for some slack, and usually more fuel can be burned to get to the target more quickly. So the multi-target solution was not always be optimal but generally it works well, if sometimes taking a few seconds to compute.

Adrian Skilling

adrian.skilling@gmail.com


ChangeLog.txt
v0.2.4
- Don't need to activate engines to enable guidance with RO (guesses ISP!)
- Doesn't shutdown engines fully in RO now (unless see below)
- Will Shutdown engines in RO for final seconds of landing to reduce thrust < gravity (like Falcon 9 does)
- Will take notice of thrust limiter setting
- Falcon 9 like descent can be enable early before engines need to be ignited

v0.2.3
- Added ability to pick multiple intermediate targets
- Removed latitude, longitude and altitude target (set target via pick target only)
- Removed max landing thrust angle, it is set to 10% of max thrust angle
- Removed acceleration gain, it is set to 4x correction factor
- Removed idle angle, it is now always 90 degrees
- Removed max thrust, this is set to 90% of the vessels maximum
- Max thrust angle in the solution is 80% of max thrust angle (allowing 20% extra for corrections)
- Min descent angle only applies to the final trajectory before the final target and not to intermediate targets
- Ground intersect constraint always applies where the ground is the lowest of all the ground targets
- The solution velocity isn't modified by the PID, which results in better following of the trajectory
- Switch to using mesh rather than LineRenderer for track and thrust vectors as transformations work better
- Golden search over time now falls back to simpler search stepping through 10 intervals between Tmin and Tmax then continues golden search, as golden search fails when probe points are non-solutions. This results in much better and faster solutions
- Fixed disappearing track bug when active for a long time

v0.2.2
- Changes to work with Realism Overhaul

v0.2.1
- Added TestCraft and TestCraftHeavy
- Tested with KSP version 1.8.1 up to 1.10.0
- Can enable/disable logging to vessel.dat and solution.dat
- Can show/hide the track, alignment vector, and thrust vectors
- Works with Realism Overhaul but trajectories look shaky


plotXYZ.py
#!/usr/bin/python

import argparse
from math import radians,sin,cos
import sys
import pylab as P
import numpy as np

def plot_line(ax,data,fx,fy,color='black',label='',linestyle='-'):
  xx=[float(d[fx]) for d in data]
  yy=[float(d[fy]) for d in data]
  ax.plot(xx,yy,color=color,label=label,linestyle=linestyle)


def plot_times(ax, times, color):
  for t in times:
    ax.axvspan(t-0.1,t+0.1,facecolor=color,alpha=0.5)

def plot_markers(ax,data,fx,fy,times,color='black',marker='o',markersize=4,alpha=1):
  cx = []
  cy = []
  for t in times:
    done = False
    for i,d in enumerate(data):
      if data[i]['time'] >= t and (not done):
        cx.append(data[i][fx])
        cy.append(data[i][fy])
        done = True
  ax.plot(cx,cy,color=color,marker=marker,markersize=markersize,linestyle='',alpha=alpha)

def plot_targets(ax,data,color='black'):
  tx = [d[0] for d in data]
  ty = [d[1] for d in data]
  ax.plot(tx,ty,color=color,marker='s',markersize=10,linestyle='')

def plot_scatter(ax,data,fx,fy,color='black'):
  tx = [d[fx] for d in data]
  ty = [d[fy] for d in data]
  ax.plot(tx,ty,color=color,marker='o',markersize=5,linestyle='')


class Vector3Time:
  def fromStr(self, s=''):
    v = s.split(":")[0]
    v = v.strip("[]")
    try:
      self.t = float(s.split(":")[1])
    except:
      self.t = -1
    self.x,self.y,self.z = [float(a) for a in v.split(",")]
    return self

  def __init__(self, x=0, y=0, z=0, t=-1):
    self.x = x
    self.y = y
    self.z = z
    self.t = t

def plot(labels,xmin,xmax,ymin,ymax,zmin,zmax,
         vxmin,vxmax,vymin,vymax,vzmin,vzmax,accelmax,tmax=30,amult=1,askip=3,
         filenames=[],showchecks=False,savepng=None,
         marktime=None,timevsfuel=False):

  # Set up figures
  # altitude against time, and throttle
  fig = P.figure(1,figsize=(15,10))

  colors=['red','blue','green','black','pink','grey','purple','salmon']

  P.subplot2grid((3,5),(0,0), colspan=1, rowspan=1)
  ax1 = P.gca()
  ax1.set_xlabel("time")
  ax1.set_ylabel("x")
  ax1.set_xlim([0,tmax])
  ax1.set_ylim([xmin,xmax])
  ax1.grid()

  P.subplot2grid((3,5),(0,1), colspan=1, rowspan=1)
  ax2 = P.gca()
  ax2.set_xlabel("time")
  ax2.set_ylabel("vx")
  ax2.set_xlim([0,tmax])
  ax2.set_ylim([vxmin,vxmax])
  ax2.grid()

  P.subplot2grid((3,5),(1,0), colspan=1, rowspan=1)
  ax3 = P.gca()
  ax3.set_xlabel("time")
  ax3.set_ylabel("y")
  ax3.set_xlim([0,tmax])
  ax3.set_ylim([ymin,ymax])
  ax3.grid()

  P.subplot2grid((3,5),(1,1), colspan=1, rowspan=1)
  ax4 = P.gca()
  ax4.set_xlabel("time")
  ax4.set_ylabel("vy")
  ax4.set_xlim([0,tmax])
  ax4.set_ylim([vymin,vymax])
  ax4.grid()

  P.subplot2grid((3,5),(2,0), colspan=1, rowspan=1)
  ax5 = P.gca()
  ax5.set_xlabel("time")
  ax5.set_ylabel("z")
  ax5.set_xlim([0,tmax])
  ax5.set_ylim([zmin,zmax])
  ax5.grid()

  P.subplot2grid((3,5),(2,1), colspan=1, rowspan=1)
  ax6 = P.gca()
  ax6.set_xlabel("time")
  ax6.set_ylabel("vz")
  ax6.set_xlim([0,tmax])
  ax6.set_ylim([vzmin,vzmax])
  ax6.grid()

  # Throttle
  if timevsfuel:
    P.subplot2grid((3,5),(0,2), colspan=1, rowspan=1)
  else:
    P.subplot2grid((3,5),(0,2), colspan=2, rowspan=1)
  ax7 = P.gca()
  ax7.set_xlabel("time")
  ax7.set_ylabel("mag(accel)")
  ax7.set_xlim([0,tmax])
  ax7.set_ylim([0,accelmax*1.1])
  ax7.grid()

  # T vs fuel
  if timevsfuel:
    P.subplot2grid((3,5),(0,3), colspan=1, rowspan=1)
    ax8 = P.gca()
    ax8.set_xlabel("time")
    ax8.set_ylabel("fuel")
    ax8.grid()
  else:
    ax8 = None

  # Attitude error
  P.subplot2grid((3,5),(0,4), colspan=1, rowspan=1)
  ax9 = P.gca()
  ax9.set_xlabel("time")
  ax9.set_ylabel("attitude error /degrees")
  ax9.set_xlim([0,tmax])
  ax9.set_ylim([0,90])
  ax9.grid()

  # XY
  P.subplot2grid((3,5),(1,2), colspan=3, rowspan=2)
  ax10 = P.gca()
  ax10.set_xlabel("x")
  ax10.set_ylabel("y")
  ax10.set_xlim([xmin,xmax])
  ax10.set_ylim([ymin,ymax])
  ax10.grid()


  for di,filename in enumerate(filenames):
    col = colors[di]
    data = read_data(filename,info)
    thrust_times = []
    check_times = []
    targets = []
    if 'minDescentAngle' in info:
      minDescentAngle = float(info['minDescentAngle'])
    else:
      minDescentAngle = None
    amin = 0
    if 'amin' in info:
      amin = float(info['amin'])
    amax = 0
    if 'amax' in info:
      amax = float(info['amax'])
    if 'thrust_times' in info:
      thrust_times = [float(t) for t in info['thrust_times'].split(",")]
    if 'check_times' in info:
      check_times = [float(t) for t in info['check_times'].split(",")]
    if not showchecks:
      check_times = []
    if 'target' in info:
      for s in info['target']:
        t=Vector3Time()
        targets.append(t.fromStr(s))
    rf = None
    if 'rf' in info:
      t=Vector3Time()
      rf = t.fromStr(info['rf'])
      targets.append(rf)
    solutions = []
    sln_Tmax = 0
    sln_fuelmin = 1000
    sln_fuelmax = 0
    if ('solution_time' in info) and ('solution_fuel' in info):
      for T,fuel in zip(info['solution_time'].split(","), info['solution_fuel'].split(",")):
        T = float(T)
        fuel = float(fuel)
        solutions.append({'T':T, 'fuel':fuel})
        sln_fuelmin = min(fuel,sln_fuelmin)
        sln_fuelmax = max(fuel,sln_fuelmax)
        sln_Tmax = max(T,sln_Tmax)

    plot_line(ax1,data,'time','x',color=col)
    plot_markers(ax1,data,'time','x',check_times,color=col)
    plot_targets(ax1,[(t.t,t.x) for t in targets])
    plot_markers(ax1,data,'time','x',[marktime],color=colors[di],markersize=10,alpha=0.5)

    plot_line(ax2,data,'time','vx',color=col)
    plot_markers(ax2,data,'time','vx',check_times,color=col)
    plot_markers(ax2,data,'time','vx',[marktime],color=colors[di],markersize=10,alpha=0.5)

    plot_line(ax3,data,'time','y',color=col)
    plot_markers(ax3,data,'time','y',check_times,color=col)
    plot_targets(ax3,[(t.t,t.y) for t in targets])
    plot_markers(ax3,data,'time','y',[marktime],color=colors[di],markersize=10,alpha=0.5)

    plot_line(ax4,data,'time','vy',color=col)
    plot_markers(ax4,data,'time','vy',check_times,color=col)
    plot_markers(ax4,data,'time','vy',[marktime],color=colors[di],markersize=10,alpha=0.5)

    plot_line(ax5,data,'time','z',color=col)
    plot_markers(ax5,data,'time','z',check_times,color=col)
    plot_targets(ax5,[(t.t,t.z) for t in targets])
    plot_markers(ax5,data,'time','z',[marktime],color=colors[di],markersize=10,alpha=0.5)

    plot_line(ax6,data,'time','vz',color=col)
    plot_markers(ax6,data,'time','vz',check_times,color=col)
    plot_markers(ax6,data,'time','vz',[marktime],color=colors[di],markersize=10,alpha=0.5)

    # plot desired magnitude of acceleration
    tdata = []
    for d in data:
      T=np.array([d['ax'],d['ay'],d['az']])
      d['mag_accel'] = np.linalg.norm(T)
    plot_line(ax7,data,'time','mag_accel',color=col)
    plot_markers(ax7,data,'time','mag_accel',check_times,color=col)
    plot_markers(ax7,data,'time','mag_accel',[marktime],color=colors[di],markersize=10,alpha=0.5)
    if 'amin' in data[0]: # continuos amin values
      plot_line(ax7,data,'time','amin',color=col,linestyle='--')
    elif amin:
      ax7.plot([0,data[-1]['time']],[amin,amin],color=col,linestyle='--')
    if 'amax' in data[0]: # continuos amax values
      plot_line(ax7,data,'time','amax',color=col,linestyle='--')
    elif amax:
      ax7.plot([0,data[-1]['time']],[amax,amax],color=col,linestyle='--')
    plot_times(ax7, thrust_times, color=col)

    if ax8:
      ax8.set_xlim([0,sln_Tmax])
      ax8.set_ylim([sln_fuelmin,sln_fuelmax])
      plot_scatter(ax8,solutions,'T','fuel',color=col)

    plot_line(ax9,data,'time','att_err',color=col)
    plot_markers(ax9,data,'time','att_err',[marktime],color=colors[di],markersize=10,alpha=0.5)


    # plot side view of X,Y
    xx,yy=[],[]
    for i,d in enumerate(data):
      xx = []
      yy = []
      xx.append(d['x'])
      yy.append(d['y'])
      xx.append(d['x']+d['ax']*amult)
      yy.append(d['y']+d['ay']*amult)
      ax10.plot(xx,yy,color=colors[di],alpha=0.5)
    plot_line(ax10,data,'x','y',color=colors[di],label=filenames[di])
    if marktime:
      plot_markers(ax10,data,'x','y',[marktime],color=colors[di],markersize=10,alpha=0.5)
    # Show checkpoints
    plot_markers(ax10,data,'x','y',check_times,color=colors[di])
    plot_targets(ax10,[(t.x,t.y) for t in targets])

  # Draw min descent angle
    if minDescentAngle is not None:
      if rf:
        fx = rf.x
        fy = rf.y
      else:
        fx = datas[0][-1]['x']
        fy = datas[0][-1]['y']
      fx = 0
      dy = 0
      s = sin(radians(minDescentAngle))
      c = cos(radians(minDescentAngle))
      d = (xmax-xmin) + (ymax-ymin)
      xx = [-d*c + fx,fx,d*c + fx]
      yy = [d*s + fy,fy,d*s + fy]
      ax10.plot(xx,yy,color=colors[di],linestyle='--')

  ax10.legend()

  fig.tight_layout(pad=0.5)
  if savepng:
    P.savefig(savepng)
  else:
    P.show()

def extract_items(line, lists=[]):
  d = {}
  for kv in line.split(" "):
    if '=' in kv:
      k,v = kv.split('=',1)
      if k not in lists:
        d[k] = v
      if k in lists:
        try:
          d[k].append(v)
        except:
          d[k] = [v]
  return d

def read_data(fname, d):
  """Reads column data file, values space or tab separated. First line in column names.
     Comments lines with hash can contain key=value pairs which will be returned in d"""
  fields=None
  dat=[]
  for line in file(fname):
    line=line.strip("\n\r")
    if line.startswith("#"):
      dd = extract_items(line[1:], lists=['target'])
      d.update(dd)
      continue
    if not fields:
      fields = line.split(None)
    else:
      try:
        data = [float(x) for x in line.split(" ")]
        if len(data)==len(fields):
          dat.append( dict(zip(fields,data)) )
      except:
        pass
  return dat

parser = argparse.ArgumentParser(description='Plot vessel data logs (or solutions) with X,Y,Z,VX,VY,VZ and Throttle in multiple plots')
parser.add_argument('filename', nargs='+',
                    help='Filename of TAB-separated data file, first line contains column names. Should contain time,x,y,z,vx,vy,vz,ax,ay,ax')
parser.add_argument('--xmin', type=float, help='Minimum x position', default=None)
parser.add_argument('--xmax', type=float, help='Maximum x position', default=None)
parser.add_argument('--ymin', type=float, help='Minimum y position', default=None)
parser.add_argument('--ymax', type=float, help='Maximum y position', default=None)
parser.add_argument('--zmin', type=float, help='Minimum z position', default=None)
parser.add_argument('--zmax', type=float, help='Maximum z position', default=None)
parser.add_argument('--vxmin', type=float, help='Minimum vx position', default=None)
parser.add_argument('--vxmax', type=float, help='Maximum vx position', default=None)
parser.add_argument('--vymin', type=float, help='Minimum vy position', default=None)
parser.add_argument('--vymax', type=float, help='Maximum vy position', default=None)
parser.add_argument('--vzmin', type=float, help='Minimum vz position', default=None)
parser.add_argument('--vzmax', type=float, help='Maximum vz position', default=None)
parser.add_argument('--marktime', type=float, help='Put a marker a this time position', default=None)
parser.add_argument('--accelmax', type=float, help='Maximum acceleration', default=None)
parser.add_argument('--tmax', type=float, help='Maximum time', default=None)
parser.add_argument('--amult', type=float, help='Multiplier for scale up thrust acceleration lines', default=1)
parser.add_argument('--square', action='store_true', help='Make XY plot square (roughly as depends on window size)', default=False)
parser.add_argument('--showchecks', action='store_true', help='Show time checks for max vel. and min descent angle', default=False)
parser.add_argument('--timevsfuel', action='store_true', help='Include plot with time vs fuel', default=False)
parser.add_argument('--savepng', help='PNG filename to save plot to', default=None)

args = parser.parse_args()

datas=[]
info={}
for filename in args.filename:
  datas.append(read_data(filename,info))
thrust_times = []
try:
  args.minDescentAngle = float(info['minDescentAngle'])
except:
  pass
try:
  thrust_times = [float(t) for t in info['thrust_times'].split(',')]
except:
  pass

alldata = []
for data in datas:
  alldata = alldata + data

# Find min and max
if not args.xmin:
  args.xmin = min([d['x'] for d in alldata])
if not args.xmax:
  args.xmax = max([d['x'] for d in alldata])
if not args.ymin:
  args.ymin = min([d['y'] for d in alldata])
if not args.ymax:
  args.ymax = max([d['y'] for d in alldata])
if not args.zmin:
  args.zmin = min([d['z'] for d in alldata])
if not args.zmax:
  args.zmax = max([d['z'] for d in alldata])

if not args.vxmin:
  args.vxmin = min([d['vx'] for d in alldata])
if not args.vxmax:
  args.vxmax = max([d['vx'] for d in alldata])
if not args.vymin:
  args.vymin = min([d['vy'] for d in alldata])
if not args.vymax:
  args.vymax = max([d['vy'] for d in alldata])
if not args.vzmin:
  args.vzmin = min([d['vz'] for d in alldata])
if not args.vzmax:
  args.vzmax = max([d['vz'] for d in alldata])
if not args.accelmax:
  args.accelmax = 0
  for d in alldata:
    T=np.array([d['ax'],d['ay'],d['az']])
    args.accelmax = max(np.linalg.norm(T),args.accelmax)

if not args.tmax:
  args.tmax = max([d['time'] for d in alldata])

if args.square:
  width = args.xmax - args.xmin
  height = args.ymax - args.ymin
  ratio = width/height
  if ratio < 1:
    args.xmin = args.xmin - (height-width)*0.5
    args.xmax = args.xmax + (height-width)*0.5
    args.zmin = args.zmin - (height-width)*0.5
    args.zmax = args.zmax + (height-width)*0.5
  if ratio > 1:
    args.ymin = args.ymin - (width-height)*0.5
    args.ymax = args.ymax + (width-height)*0.5

plot(args.filename,xmin=args.xmin,xmax=args.xmax,ymin=args.ymin,ymax=args.ymax,zmin=args.zmin,zmax=args.zmax,
     vxmin=args.vxmin,vxmax=args.vxmax,vymin=args.vymin,vymax=args.vymax,vzmin=args.vzmin,vzmax=args.vzmax,tmax=args.tmax,
     amult=args.amult,filenames=args.filename,showchecks=args.showchecks,accelmax=args.accelmax,savepng=args.savepng,
     marktime=args.marktime, timevsfuel=args.timevsfuel)
